### PR TITLE
Fix SIGNEXTEND EVM issue and harden execution layer

### LINKED ISSUE #696: Burrow's EVM returns results different from Remix(Javascript VM)
Please include in your bug report:

- [0.17.1 ] `burrow version` (docker image tag or branch if built from source)
- [ No monax] `monax version` (if applicable)
- [1.9.4 ] `go version` (if applicable)
- [No ducker ] `docker version` (if applicable)
- [Ubuntu 16.04 amd64 ] operating system details (osx/windows/linux)

- [ ] steps to reproduction

Burrow's EVM issue :
I did deployed this smart contract on the Burrow, and called the Test method with this parameters 5,6,7 
but the result is very different with Remix Ide (Ethereum JavaScript VM)
Here is the issue information :
*******************************************************************
Smart Contract :

```solidity
pragma solidity ^0.4.16;


contract ByteCasting {
    

      function Test(int8 _in1, int256 _in2, int16 _in3) public pure 
	  			returns(bytes out,int8 _out1, int256 _out2, int16 _out3) {

        bytes memory _buff = new  bytes(128);

        //////////////
        // Serializing
        uint128 _offst = 128;

        int8ToBytes(_offst,_in1,_buff);
        _offst -= 1;
        
        int256ToBytes(_offst,_in2,_buff);
        _offst -= 32;        
        
        int16ToBytes(_offst,_in3,_buff);
        _offst -= 2; 
        
        out = _buff;
        ////////////////
        // Deserializing
        _offst = 128;
    
        _out1 = bytesToInt8(_offst,_buff);
        _offst -= 1;
        
        _out2 = bytesToInt256(_offst,_buff);
        _offst -= 32;
        
        _out3 = bytesToInt16(_offst,_buff);
        _offst -= 2;

    }
    

    function int8ToBytes(uint128 _offst, int8 _input, bytes _output) public pure {

        assembly {
            mstore(add(_output, _offst), _input)
        }
    }
    
    function int16ToBytes(uint128 _offst, int16 _input, bytes _output) public pure {

        assembly {
            mstore(add(_output, _offst), _input)
        }
    }

    function int256ToBytes(uint128 _offst, int256 _input, bytes _output) public pure {

        assembly {
            mstore(add(_output, _offst), _input)
        }
    }
    

    function bytesToInt8(uint128 _offst, bytes _input) public pure returns (int8 _output) {
        
        assembly {
            _output := mload(add(_input, _offst))
        }
    }
    
    function bytesToInt16(uint128 _offst, bytes _input) public pure returns (int16 _output) {
        
        assembly {
            _output := mload(add(_input, _offst))
        }
    }

    function bytesToInt256(uint128 _offst, bytes _input) public pure returns (int256 _output) {
        
        assembly {
            _output := mload(add(_input, _offst))
        }
    }
}
```

*******************************************************************
Function call :

Test(5,6,7)

*******************************************************************
what actually happened:
Burrow's output :+1: 
[Burrow-EVM-Issue.txt](https://github.com/hyperledger/burrow/files/1759021/Burrow-EVM-Issue.txt)

"0000000000000000000000000000000000000000000000000000000000000000
 0000000000000000000000000000000000000000000000000000000000000000
 0000000000000000000000000000000000000000000000000000000000800600
 0000000000000000000000000000000000000000000000000000000000000684",
"-124",
"6",
"-32762"

*******************************************************************
-  what you expected to happen
Remix(Ethereum) output :

 0000000000000000000000000000000000000000000000000000000000000000
 0000000000000000000000000000000000000000000000000000000000000000 
 0000000000000000000000000000000000000000000000000000000000000700
 0000000000000000000000000000000000000000000000000000000000000605",
	"1": "int8: 	_out1 5",
	"2": "int256: 	_out2 6",
	"3": "int16: 	_out3 7"

### PR DESCRIPTION
**Pull request by silasdavis**
This fixes an implementation bug in the `SIGNEXTEND` (https://github.com/hyperledger/burrow/blob/master/manager/burrow-mint/evm/vm.go#L355-L372) where the mask was added rather than anded: `numb.Add(numb, mask)` when it should have been: `numb.And(numb, mask)`

In fixing this I took the opportunity to centralise our logic for using `big.Int`, clean up our arithmetic code, and do a manual audit of other opcodes to see if there were any other issues on inspection. I didn't find anything else so the refactored code should function in a mostly identical way (we have some added safety around overflow by the use of big ints. This fixes #696. I have a full integration test coming in Bosmarmot.

This also removes a panic from memory fixing #697 and uses `big.Int` for bounds checking and future expansion (e.g. paging memory).

It also stops panics caused by a tx triggering a bug from taking down a chain by recovering from the panic and returning an error.

### CODE REVIEW

File: execution/evm/memory.go [hunk 1]
```diff
@@ -54,10 +58,22 @@ func DefaultDynamicMemoryProvider() Memory {
...
+func (mem *dynamicMemory) Read(offset, length *big.Int) ([]byte, error) {
+	// Ensures positive and not too wide
+	if !offset.IsUint64() {
+		return nil, fmt.Errorf("offset %v does not fit inside an unsigned 64-bit integer", offset)
+	}
+	// Ensures positive and not too wide
+	if !length.IsUint64() {
+		return nil, fmt.Errorf("length %v does not fit inside an unsigned 64-bit integer", offset)
+	}
+	return mem.read(offset.Uint64(), length.Uint64())
```
Comment (silasdavis) [inline execution/evm/memory.go]: For now I am wrapping the implementation and using big int as a convenient way to check bounds.

**Comment by compleatang** [comment]
Fipf reviewed.

**Comment by dennismckinnon** [comment]
Looks good to me

**Comment by pouladzade** [comment]
I reviewed it, It seems ok

**Comment by smblucker** [comment]
LGTM

**Pull request merged**